File: CaseStudies/noPCM/src/CSharp/Calculations.cs
using System.Collections.Generic;
using System.Linq;
using Microsoft.Research.Oslo;

public class Calculations {
    public static List<double> func_T_W(InputParameters inParams) {
        Vector f(double t, Vector T_W) {
          return new Vector((1/inParams.tau_W) * (inParams.T_C - T_W[0]));
        }

        Options opts = new Options();
        opts.AbsoluteTolerance = inParams.A_tol;
        opts.RelativeTolerance = inParams.R_tol;

        IEnumerable<SolPoint> sol = Ode.GearBDF(
            0, 
            new Vector(inParams.T_init), 
            f,
            opts);

        IEnumerable<SolPoint> pts = sol.SolveFromToStep(0, inParams.t_final, inParams.t_step);

        List<double> T_W = new List<double>();
        foreach (SolPoint sp in pts) {
          T_W.Add(sp.X[0]);
        }

        return T_W;
    }
}

File: CaseStudies/noPCM/src/Cpp/Calculations.cpp
#include "Calculations.hpp"

#include <vector>

#include "boost/numeric/odeint.hpp"
#include "InputParameters.hpp"
#include "NoPCMODE.hpp"

using std::vector;

vector<double> func_T_W(InputParameters inParams) {
  NoPCMODE ode = NoPCMODE(inParams.T_C, inParams.tau_W);
  vector<double> T_W;
  vector<double> init {inParams.T_init};
  boost::numeric::odeint::runge_kutta_dopri5<vector<double>> rk = boost::numeric::odeint::runge_kutta_dopri5<vector<double>>();
  auto stepper = boost::numeric::odeint::make_controlled(inParams.A_tol, inParams.R_tol, rk);
  boost::numeric::odeint::integrate_const(stepper, ode, init, 0.0, inParams.t_final, inParams.t_step, Populate_T_W(T_W));

  return T_W;
}